x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/ShardMultiSearchAction.java

```diff
@@ -0,0 +1,244 @@
+
+        public Request(MultiSearchRequest multiSearchRequest) {
+            super(multiSearchRequest.requests().get(0).indices()[0]);
+            this.multiSearchRequest = multiSearchRequest;
+            assert multiSearchRequest.requests().stream()
+                .map(SearchRequest::indices)
+                .flatMap(Arrays::stream)
+                .distinct()
+                .count() == 1 : "action [" + NAME + "] cannot handle msearch request pointing to multiple indices";
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            multiSearchRequest = new MultiSearchRequest();
+            multiSearchRequest.readFrom(in);
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = validateNonNullIndex();
+            if (index.startsWith(EnrichPolicy.ENRICH_INDEX_NAME_BASE) == false) {
+                validationException = ValidateActions.addValidationError("index [" + index + "] is not an enrich index",
+                    validationException);
+            }
+            return validationException;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            multiSearchRequest.writeTo(out);
+        }
+
+        MultiSearchRequest getMultiSearchRequest() {
+            return multiSearchRequest;
+        }
+    }
+
+    public static class TransportAction extends TransportSingleShardAction<Request, MultiSearchResponse> {
+
+        private final IndicesService indicesService;
+
+        @Inject
+        public TransportAction(ThreadPool threadPool, ClusterService clusterService, TransportService transportService,
+                               ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                               IndicesService indicesService) {
+            super(NAME, threadPool, clusterService, transportService, actionFilters, indexNameExpressionResolver,
+                Request::new, ThreadPool.Names.SEARCH);
+            this.indicesService = indicesService;
+        }
+
+        @Override
+        protected Writeable.Reader<MultiSearchResponse> getResponseReader() {
+            return MultiSearchResponse::new;
+        }
+
+        @Override
+        protected boolean resolveIndex(Request request) {
+            return true;
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            String index = request.concreteIndex();
+            IndexRoutingTable indexRouting = state.routingTable().index(index);
+            int numShards = indexRouting.shards().size();
+            if (numShards != 1) {
+                throw new IllegalStateException("index [" + index + "] should have 1 shard, but has " + numShards + " shards");
```

Comment:
Are there are convenient post index settings PUT hooks that we can help to alert the user earlier that .enrich-* can not be changed from 1 shard ?

Reply:
The number of primary shards is an immutable setting, so if a user tries to change that on an enrich index then that will fail. I guess this is more of an assertion than user validation. For example if we by mistake query other indices.